nexus/db-queries/src/db/datastore/target_release.rs

```diff
@@ -0,0 +1,258 @@
+use crate::db::error::{public_error_from_diesel, ErrorHandler};
+use crate::db::model::{TargetRelease, TargetReleaseSource};
+use crate::db::schema::target_release::dsl;
+use crate::transaction_retry::OptionalError;
+use async_bb8_diesel::AsyncRunQueryDsl as _;
+use diesel::insert_into;
+use diesel::prelude::*;
+use omicron_common::api::external::{CreateResult, LookupResult};
+
+impl DataStore {
+    /// Fetch the current target release, i.e., the row with the largest
+    /// generation number.
+    pub async fn get_target_release(
+        &self,
+        opctx: &OpContext,
+    ) -> LookupResult<TargetRelease> {
+        opctx.authorize(authz::Action::Read, &authz::FLEET).await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+
+        // Fetch the row in the `target_release` table with the largest
+        // generation number. The subquery accesses the same table, so we
+        // have to make an alias to not confuse diesel.
+        let target_release2 = diesel::alias!(
+            crate::db::schema::target_release as target_release_2
+        );
+        dsl::target_release
+            .filter(dsl::generation.nullable().eq_any(target_release2.select(
+                diesel::dsl::max(target_release2.field(dsl::generation)),
+            )))
+            .select(TargetRelease::as_select())
+            .first_async(&*conn)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// Insert a new target release row and return it. It will only become
+    /// the current target release if its generation is larger than any
+    /// existing row.
+    pub async fn set_target_release(
+        &self,
+        opctx: &OpContext,
+        target_release: TargetRelease,
+    ) -> CreateResult<TargetRelease> {
+        opctx.authorize(authz::Action::Modify, &authz::FLEET).await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+        let err = OptionalError::new();
+        self.transaction_retry_wrapper("set_target_release")
+            .transaction(&conn, |conn| {
+                let target_release = target_release.clone();
+                let err = err.clone();
+                async move {
+                    // Ensure that we have a TUF repo representing a system version.
+                    if let TargetReleaseSource::SystemVersion =
+                        target_release.release_source
+                    {
+                        if let Some(system_version) =
+                            target_release.system_version.clone()
+                        {
+                            assert_eq!(
+                                system_version.clone(),
+                                self.update_tuf_repo_get(opctx, system_version)
+                                    .await
+                                    .map_err(|e| err.bail(e))?
+                                    .repo
+                                    .system_version,
```

Comment:
We don't want to panic here, right?

If we were doing this in raw SQL I might do:

```sql
INSERT INTO target_release (id, ...) VALUES (SELECT id FROM tuf_repo WHERE system_version = ...), ...)
```

This would avoid an interactive transaction and reduce contention a lot but I guess it's fairly painful to do in Diesel.  We could also fetch the id outside of the transaction but I suppose we still have to check that it's still there so it doesn't help much.

Reply:
Moved this resolution to the app layer in 80f69029c.